torch/lib/c10d/test/ProcessGroupNCCLTest.cpp

```diff
@@ -0,0 +1,236 @@
+
+  std::shared_ptr<c10d::ProcessGroup::Work> run(int rootRank, int rootTensor) {
+    // For the duration of this function, make THC use our streams
+    auto guards = createStreamGuard();
+
+    // Launch sleep on every device
+    for (auto i = 0; i < numDevices_; i++) {
+      CUDADevice device(i);
+      cudaSleep(streams_[i], 2000 * 1000 * 1000);
+    }
+
+    // Launch value initialization for every tensor
+    for (auto i = 0; i < numDevices_; i++) {
+      CUDADevice device(i);
+      inputs_[i].fill_(pg_->getRank() * numDevices_ + i);
+    }
+
+    ::c10d::BroadcastOptions options;
+    options.rootRank = rootRank;
+    options.rootTensor = rootTensor;
+    return pg_->broadcast(inputs_, options);
+  }
+};
+
+void testAllreduce(const std::string& path, int rank, int size) {
+  auto test = AllreduceNCCLTest(path);
+  test.initialize(rank, size);
+  auto work = test.run();
+  // Wait for work to finish
+  test.wait(work);
+
+  // Validation
+  const int totalNumGPUs = test.numDevices() * size;
+  const auto expected = (totalNumGPUs * (totalNumGPUs - 1)) / 2;
+  auto tensors = test.getTensors();
+  for (auto j = 0; j < tensors.size(); j++) {
+    auto& tensor = tensors[j];
+    auto data = tensor.data<float>();
+    for (auto k = 0; k < tensor.numel(); k++) {
+      if (data[k] != expected) {
+        throw std::runtime_error("BOOM!");
+      }
+    }
+  }
+  std::cout << "Allreduce test successful" << std::endl;
+}
+
+void testBroadcast(const std::string& path, int rank, int size) {
+  auto test = BroadcastNCCLTest(path);
+  test.initialize(rank, size);
+
+  const int numDevices = test.numDevices();
+  // Try every permutation of root rank and root tensor
+  for (auto rootRank = 0; rootRank < size; rootRank++) {
+    for (auto rootTensor = 0; rootTensor < numDevices; rootTensor++) {
+      auto work = test.run(rootRank, rootTensor);
+
+      // Wait for work to complete
+      test.wait(work);
+
+      // Check results
+      const auto expected = (rootRank * numDevices + rootTensor);
+      auto tensors = test.getTensors();
+      for (auto j = 0; j < tensors.size(); j++) {
+        auto& tensor = tensors[j];
+        auto data = tensor.data<float>();
+        for (auto k = 0; k < tensor.numel(); k++) {
+          if (data[k] != expected) {
+            throw std::runtime_error("BOOM!");
+          }
+        }
+      }
+    }
+  }
+  std::cout << "Broadcast test successful" << std::endl;
+}
+
+int main(int argc, char** argv) {
+  // Use WORLD_SIZE and RANK environmental variables to do multi-node
+  // distributed testing
+  auto sizeEnv = std::getenv("WORLD_SIZE");
+  auto rankEnv = std::getenv("RANK");
+
+  int size = 1;
+  int rank = 0;
+
+  if (sizeEnv && rankEnv) {
+    size = std::stoi(std::string(sizeEnv));
+    rank = std::stoi(std::string(rankEnv));
+    std::cout << "Multi-node world size: " << size << " rank: " << rank
+              << std::endl;
+  }
+
+  TemporaryFile file;
```

Comment:
The different tests are not scoped to the store, so they will end up writing to the same keys in this file. To fix this you can either use a different file per test, or add some kind of scoped store (there is one in caffe2/distributed for the caffe2 store API).